docs/reference/builder.md

```diff
@@ -953,6 +953,41 @@ For example:
 The output of the final `pwd` command in this `Dockerfile` would be
 `/path/$DIRNAME`
 
+## ARG
+    ARG [--default=<value>] [--description=<value>] <arg name>
+
+The `ARG` instruction defines a variable that is allowed to be passed as
+build-time variable to the builder. It is used to define a single variable at a time.
+Multiple variables may be defined by specifying `ARG` more than once. The `ARG`
+instruction can optionally take default value and description of the variable
+being defined as well. These are persisted in the image history so they can be
+useful for documentation. Additionally, the default value of the variable is used
+for expansion/substitution or passing to the `RUN` command when a build-time
+variable is not passed to the builder.
+
+A variable definition comes into effect from the line it appears in the `Dockerfile`
+i.e. the build-time values for that variable will only be available for
+expansion/substitution beginning from that line. For example:
+
+    $ docker build --build-arg user=foo1 <<MARK
+     -> ...
+    l-> USER ${user:-foo}
+    m-> ARG $user
+    n-> USER $user
```

Comment:
They syntax here wasn't obvious to me. Perhaps just make it:
```
$ docker build --build-arg user=foo1 <<EOF
USER ${user:-foo}
ARG $user
USER $user
EOF
```
and then just refer to the lines by number. E.g. `The USER at line 1.... The USER one line 3....`